### ISSUE TITLE
About PDO JSON support

### ISSUE THREAD

**Issue by fengqi** [perftools/xhgui#439]
I want to provide special support based on the drive name, for example, add `json_support` switch in the PDO config

different db JSON syntax will be different, sql may also need to write multiple

known and commonly used mysql >=5.7, postgresql >=9.2, sql server >=2016

**Comment by glensc** [comment]
I added the driver name to PdoRepository:
- https://github.com/perftools/xhgui/pull/438

for this merge request:
- https://github.com/perftools/xhgui/pull/437

(...but that author disappeared)

anyway, now driver-specific code is now possible.

this must be enabled in config:

```php
"enable_json" => true // (or something similar)
```

as even if your db server supports json, you may still may have old schema in place

**Comment by glensc** [comment]
also, depending on how invasive the changes are, maybe need another `PdoRepository`, like `PdoJsonRepository`...

**Comment by glensc** [comment]
Also, when you start working on it, try to make atomic commits:
- https://www.freshconsulting.com/insights/blog/atomic-commits/
- https://en.wikipedia.org/wiki/Atomic_commit#Atomic_commit_convention

makes much easier to rebase, do codereview and apply fixes

(should put this to CONTRIBUTING.md)

**Comment by fengqi** [comment]
> also, depending on how invasive the changes are, maybe need another `PdoRepository`, like `PdoJsonRepository`...

Yes, our internal current implementation is for mysql, there is a file `PdoMySQLRepository` that overrite the `buildWhere` and `initSchema` method

**Comment by fengqi** [comment]
> Also, when you start working on it, try to make atomic commits:
> 
> * https://www.freshconsulting.com/insights/blog/atomic-commits/
> * https://en.wikipedia.org/wiki/Atomic_commit#Atomic_commit_convention
> 
> makes much easier to rebase, do codereview and apply fixes
> 
> (should put this to CONTRIBUTING.md)

ok

**Comment by glensc** [comment]
> > also, depending on how invasive the changes are, maybe need another `PdoRepository`, like `PdoJsonRepository`...
> 
> Yes, our internal current implementation is for mysql, there is a file `PdoMySQLRepository` that overrite the `buildWhere` and `initSchema` method

I prefer [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance), but anyway, perhaps you submit your code and then I can then refactor to composition and provide fallback to other drivers.

**Comment by fengqi** [comment]
> > > also, depending on how invasive the changes are, maybe need another `PdoRepository`, like `PdoJsonRepository`...
> > 
> > 
> > Yes, our internal current implementation is for mysql, there is a file `PdoMySQLRepository` that overrite the `buildWhere` and `initSchema` method
> 
> I prefer [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance), but anyway, perhaps you submit your code and then I can then refactor to composition and provide fallback to other drivers.

you mean like trait

**Comment by glensc** [comment]
> you mean like trait

no. but don't bother if you don't understand, I'll take the refactoring on me.

**Comment by fengqi** [comment]
> > you mean like trait
> 
> no. but don't bother if you don't understand, I'll take the refactoring on me.

ok

**Comment by glensc** [comment]
@fengqi seems you have not pushed your code yet

**Comment by fengqi** [comment]
@glensc  sry, our focus is on migrating to the GoLang project recently, so i haven’t had time yet.
maybe i can temporarily close this issue and open it again later.

**Comment by glensc** [comment]
@fengqi but i only asked you push your code to some pull request, and can continue from there.

**Comment by fengqi** [comment]
> @fengqi but i only asked you push your code to some pull request, and can continue from there.

ok

**Comment by glensc** [comment]
@fengqi just push the changes and share the branch.
